apache/superset
# feat: Stop pending queries when user close dashboard

### SUMMARY
This is a feature requested from airbnb. When user open a large dashboard and apply filters, very possibly there is no cached query results, so the dashboard will send queries to Presto, some of them might be very slow. In airbnb, Presto resource is allocated by user. Each user can run 3 concurrent queries at a time, and extra queries are queued up. As the result new queries will not be executed until queued queries were finished first.

This limitation of using Presto resource sometime causes bad experience for Superset dashboard users. Once user's queries are queued up, there is no way to free up, and users have to wait pretty long time to back to normal (after all the previous queued queries are completed). Both users and Superset have no way to stop the queued queries, even users may want to abandon them.

To avoid this situation, this PR proposes a solution: Superset dashboard detects **browser close event**. When it happens, Dashboard will call a Presto API, with username and dashboard_id, then Presto will kill/stop all the pending and waiting queries from this user and dashboard.

**Note**:
This solution depends on an airbnb **internal API call**, which is **not** available for other corporations. This PR is to provide a hook that connects Dashboard to internal API so that we can expands Superset functionalities.

### TEST PLAN
1. Open Superset dashboard, and open presto web interface to watch for new queries
2. Apply dashboard filter, should see new queries started in Presto
3. Close dashboard from browser, or open other url in the same browser window.
4. Should see pending queries are killed by Presto.

cc @etr2460 @john-bodley @ktmud
